### DISCUSSION TITLE
How RTDN pull messages feature work?

### DISCUSSION THREAD

**Discussion by ciao-chung** [imdhemy/laravel-in-app-purchases#19] [Q&A] [answered]
After reading [official document](https://developer.android.com/google/play/billing/getting-ready#configure-rtdn) about RTDN  and [offical gcloud pub/sub php library](https://github.com/googleapis/google-cloud-php/tree/master/PubSub#sample).

As far as I know, I must set up **subscription** of gcloud pub/sub then I can pull messages.

But in this package I can't find configuration about gcloud **subscription**.

So I curious how this package work for pull.

**Comment by imdhemy** [comment, marked as answer]
Hi, @ciao-chung!

The process is much simpler. Your code **shouldn't** depend on the **pull** mechanism due to:
- Inefficient use of the Google Play Developer API can lead to API quota restrictions.
- You must handle the events yourself, through a cron job or any other appropriate way.

**Otherwise**, depending on the **push mechanism** is much easier and efficient. The push mechanism delegates handling events to Google, just wait and Google will notify your server about all the subscription updates.

**All that you need is to follow the following steps:**

1. Once the subscription is purchased in your mobile app, send a request to your backend containing the subscription base64_encoded token.
2. This token can be parsed and validated using the `Imdhemy\Purchases\Facades\Subscription` facade. The completion of the request and get a response means a valid token. It's up to you to maintain the uniqueness of the purchase tokens.

```php
$subscriptionReceipt = Subscription::googlePlay()->id($itemId)->token($token)->get();
```
check [this section](https://github.com/imdhemy/laravel-in-app-purchases#google-play-subscriptions) for more information.

3. After setting up your RTDN to be pushed to the [provided URI](https://github.com/imdhemy/laravel-in-app-purchases#i1-routing) `https://yoursite.example/purchases/subscriptions/google`, the proper event will be triggered automatically once a notification is received from Google, just implement a listener for the event you want to handle.
4. Any [triggered event](https://github.com/imdhemy/laravel-in-app-purchases#purchase-events) provides the same collection of the required method to handle the subscription update. Check [this section](https://github.com/imdhemy/laravel-in-app-purchases#purchase-events) for more information.

I hope this illustrates the full process.

**Reply by ciao-chung** [reply]
Hi @imdhemy .
Thanks a lot for your description.
It is very helpful to me.
Now I finally understand the whole structure.
Thanks for your help again.
Have a nice day, bro.
😎

**Reply by imdhemy** [reply]
You are welcome! ^^

**Reply by SlavaSt295** [reply]
@imdhemy im sorry for this 100% stupid question, but i cant find solution for many days...

"3. After setting up your RTDN to be pushed to the provided URI" 
Where in google console i can set up this endpoint uri of my server ?) 

![image](https://user-images.githubusercontent.com/38761838/115786809-fa6af680-a3c9-11eb-988f-16bbcba1740e.png)
Following documentation i already set up topic and subscription. when i click pull button, i got messages. But i really dont understand how can i sent thems to the server)

Hope you undestand me)

**Reply by SlavaSt295** [reply]
OMG, sorry for that. Found it 🤪
Your product is awesome. Best wishes!

![image](https://user-images.githubusercontent.com/38761838/115836204-e05c0300-a41f-11eb-92fe-4f4759d893a7.png)

**Comment by jbagaresgaray** [comment]
Hi @ciao-chung @imdhemy .... is there a way where we could test the RTDN locally? I just want to see the details inside this each events [https://github.com/imdhemy/laravel-in-app-purchases#purchase-events](https://github.com/imdhemy/laravel-in-app-purchases#purchase-events). I just want to make sure and validate if my Pub/Sub really works

**Reply by imdhemy** [reply]
@jbagaresgaray For Google play you can send a test notification and check the logs to find it there.
